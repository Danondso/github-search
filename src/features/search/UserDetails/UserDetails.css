.user-details-container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    border-radius: 15px;
    background-color: #eee8aa;
    border-radius: 15px;
}

.user-details-item {

  font-size: smaller;
  padding: 0.5rem;
  color: #282c34;
  text-overflow: ellipsis;
}

.user-details-row {
    display: flex;
    justify-content: center;
}
  
