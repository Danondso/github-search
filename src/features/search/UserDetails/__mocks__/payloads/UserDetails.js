// cutting down to what's needed for the test
export const userDetailsPayload = {
  bio: 'Beep beep boop',
  twitter_username: null,
  public_repos: 41,
  public_gists: 1,
  followers: 11,
  following: 27,
  created_at: '2014-03-20T19:22:59Z',
  updated_at: '2022-01-29T15:07:30Z',
};

// truncating this payload to only what we need and a tiny bit of context info (id and name),
export const userDetailsReposPayload = [
  {
    id: 443429558,
    name: 'bedfellow',
  },
  {
    stargazers_count: 34,
  },
  {
    stargazers_count: 0,
  },
];
