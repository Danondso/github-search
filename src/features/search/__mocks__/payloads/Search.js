export default {
  total_count: 3,
  incomplete_results: false,
  items: [
    {
      login: 'Danondso',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },
    {
      login: 'GerardWay',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },
    {
      login: 'RayToro',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },

  ],
};
