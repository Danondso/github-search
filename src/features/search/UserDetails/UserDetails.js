import React, { useEffect, useState } from 'react';
import axios from 'axios';
import propTypes from 'prop-types';
import getAuthorizationHeaders from '../../../utils/authorizationUtil/authorizationUtil';
import './UserDetails.css';

const authHeaders = getAuthorizationHeaders();
const baseUrl = `${process.env.REACT_APP_GITHUB_API_BASE_URL}/users/`;

function UserDetails({ userName }) {
  const [userDetails, setUserDetails] = useState(undefined);
  const [userDetailsError, setUserDetailsError] = useState(undefined);

  useEffect(async () => {
    // normally I'd put the other call within a function, but since we're leveraging the second call
    // immediately I've opted to keep it here since we enrich user details with it.
    try {
      const { data } = await axios.get(`${baseUrl}${userName}`, authHeaders);
      const userRepoResponse = await axios.get(`${baseUrl}${userName}/repos`, authHeaders);
      setUserDetails({
        ...data,
        // short-circuit on reduce and set to zero if we can't reduce the data
        star_count: userRepoResponse.data?.reduce((total, number) => total + number, 0) || 0,
      });
    } catch (error) {
      setUserDetailsError(error.message);
    }
  }, [userName]);

  const renderRow = (children) => (
    <div className="user-details-item user-details-row">
      {children}
    </div>
  );

  return (
    <div className="user-details-container">
      {userDetails && !userDetailsError && (
        <div className="user-details-item">
            {userDetails.bio && renderRow(`Bio: ${userDetails.bio}`)}
            {userDetails.company && renderRow(`Company: ${userDetails.company}`)}
            {userDetails.blog && renderRow(`Blog: ${userDetails.blog}`)}
            {userDetails.followers && renderRow(`Followers: ${userDetails.followers}`)}
            {userDetails.public_repos && renderRow(`Public Repos: ${userDetails.public_repos}`)}
            {userDetails.star_count && renderRow(`Star Count: ${userDetails.public_repos}`)}
        </div>
      )}

      {userDetailsError && (
      <div className="user-details-item ">
        {`Error: ${userDetailsError}`}
      </div>
      )}
    </div>

  );
}

UserDetails.propTypes = {
  userName: propTypes.string.isRequired,
};

export default UserDetails;
