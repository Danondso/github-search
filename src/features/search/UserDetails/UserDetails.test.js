import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import { act } from '@testing-library/react';
import UserDetails from './UserDetails';

import { userDetailsPayload, userDetailsReposPayload } from './__mocks__/payloads/UserDetails';

jest.mock('axios');

describe('UserDetails Test Suite', () => {
  // the entire container setup to render this is lifted from https://reactjs.org/docs/testing-recipes.html#act
  let container;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  test('Renders with required props', async () => {
    axios.get.mockReturnValueOnce({ data: userDetailsPayload });
    axios.get.mockReturnValueOnce({ data: userDetailsReposPayload });
    await act(async () => {
      ReactDOM.render(<UserDetails
        userUrl="http://locahost"
        reposUrl="http://locahost"
      />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('On Error, the user details displays the error', async () => {
    axios.get.mockReturnValueOnce(Promise.resolve(userDetailsPayload));
    axios.get.mockRejectedValueOnce({ message: 'Something awful happened upstairs' });
    await act(async () => {
      ReactDOM.render(<UserDetails
        userUrl="http://locahost"
        reposUrl="http://locahost"
      />, container);
    });
    expect(container).toMatchSnapshot();
  });
});
