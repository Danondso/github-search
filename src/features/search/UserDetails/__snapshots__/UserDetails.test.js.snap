// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Test Suite On Error, the user details displays the error 1`] = `
<div>
  <div
    class="user-details-container"
  >
    <div
      class="user-details-item "
    >
      Error: Something awful happened upstairs
    </div>
  </div>
</div>
`;

exports[`UserDetails Test Suite Renders with required props 1`] = `
<div>
  <div
    class="user-details-container"
  >
    <div
      class="user-details-item"
    >
      <div
        class="user-details-item user-details-row"
      >
        Bio: Beep beep boop
      </div>
      <div
        class="user-details-item user-details-row"
      >
        Followers: 11
      </div>
      <div
        class="user-details-item user-details-row"
      >
        Public Repos: 41
      </div>
      <div
        class="user-details-item user-details-row"
      >
        Star Count: 41
      </div>
    </div>
  </div>
</div>
`;
